### ISSUE TITLE
Add full W3C Trace Context support

### ISSUE THREAD

**Issue by danielkhan** [getsentry/team-sdks#41]
### Description

While we are moving to a world where we may process, propagate, and ingest pure OpenTelemetry (OTLP) data, we have to reconsider our trace context propagation schema and adopt [W3C Trace Context](https://www.w3.org/TR/trace-context/) fully. 

## About W3C Trace Context
W3C Trace Context is a standard for trace context propagation.
It defines three headers.
* `trace-parent` to capture the relationship between spans and propagate sampling decisions
* `trace-state`: A _vendor-prefixed_ key-value list to hold vendor-specific (meta) data, like tenant IDs or vendor-local trace IDs
* `baggage`: A userland read- and writeable key-value list of properties that can propagate arbitrary information—E.g., for a/b tests or for propagating a customer loyalty status. 

Sentry has a very similar trace propagation concept.
It defines
* `sentry-trace`, which is almost similar to `trace-state`, except missing some flags W3C introduced
* `baggage` where we follow the spec, *yet* we are using it to propagate essential sentry-specific metadata like the dynamic sampling context (DSC)

We are not using a concept like `trace-state`. 

## Problem Description
The problem arises in mixed scenarios where one tier might be instrumented purely with OpenTelemetry.

![Mixed Trace scenario](https://github.com/getsentry/team-sdks/assets/2722161/3b525462-38fb-42b5-8b4d-960bea00138a)

In this example, the web frontend sends sentry trace headers to the Node middleware where no Sentry SDK or propagator is present, which leads to context loss ( a broken trace).

## Out Of Scope
While we are currently not using the `baggage` header to the spec, this problem is out-of-scope for this project.

## Next steps
* We must find a migration path that lets us first emit `sentry-trace` and `trace-parent` in parallel. This can't be turned on by default because of `CORS,` though
* Later, we should come to a world where we fully support Trace Context and deprecate `sentry-trace`

## Time constraints
We will soon be able to ingest OTel traces and this turns out to be rather critical. So we should have a plan by end of Q4 and implement this in Q1.

### Stakeholder(s)

* @stephanie-anderson 
* @smeubank 
* @AbhiPrasad 
* @phacops
* @jjbayer

### Team(s)

Web-Backend, Mobile, Web-Frontend

**Comment by cleptric** [comment]
> We must find a migration path that lets us first emit sentry-trace and trace-parent in parallel. This can't be turned on by default because of CORS, though

This is only a concern in FE JS, all other platforms can in theory send both headers today.

> We also need to ensure that we can pick up an incoming trace-parent header

The question will be what will take precedence for the moment, likely `sentry-trace`.

**Comment by danielkhan** [comment]
> The question will be what will take precedence for the moment, likely sentry-trace

Yes, `sentry-trace` it is.

**Comment by krystofwoldrich** [comment]
React-Native will get this from JS.

**Comment by ericsampson** [comment]
Hi! I'mvery  glad to see this being worked on.

We have a large mixed OTel-and-Sentry microservice system. Is there any docs/guidance on how we can get end-to-end distributed tracing working? (doesn't have to be polished, even just links to GH issues or pointers would be helpful)

I'm also wondering what direction we should take re evolving our current setup, in terms of matching the direction that Sentry is going... Should we be adding the Sentry SDK to the systems that are currently instrumented with Otel, or the reverse and moving to a "fully-OTel but with Sentry exporter" system.

The sense I'm getting is that the Sentry OTel story is still evolving, and so it may be too early for us to move to the latter direction without experiencing a ton of "bleeding-edge paper cuts" ? But at the same time I don't really feeling like doing two migrations (go all-Sentry now and then Sentry-on-OTel later) since there are so many micro services that need touching.

The trace header interior/propagation story is kinda messy in a legacy mixed system like ours. And for more fun, we also have legacy Honeycomb usage (with their proprietary headers) in the mix for some of the services, so there are 3 standards in play 🥴 

Thanks!

**Comment by danielkhan** [comment]
Hi @ericsampson, the direct answer is switching to Sentry SDKs is more beneficial. We're planning to support OTLP ingest this year, but OTel's data often doesn't match Sentry's quality, particularly for errors.

I suggest using our SDKs as we're integrating OTel for its valuable instrumentation. For instance, our Node.js [experimental package](https://www.npmjs.com/package/@sentry/node-experimental) already utilizes OTel for auto-instrumentation while maintaining compatibility with the OTel tracing SDK and providing Sentry's error tracking.

Throughout this year, we'll introduce more SDKs that integrate OTel. If installing our SDK isn't an option, you'll still be able to send data via OTLP, which is part of the improvements we're discussing here.

**Comment by ericsampson** [comment]
thanks for the info @danielkhan !

**Comment by brustolin** [comment]
Snippet to generate `traceparent` with the Cocoa SDKs.
```swift
func getTraceParent() -> String? {
        guard let span = SentrySDK.span else { return nil }
        return "00-\(span.traceId.sentryIdString)-\(span.spanId.sentrySpanIdString)-0\(span.sampled == .yes ? 1 : 0)"
    }
```
